src/bsoncxx/lib/bsoncxx/v1/types/value.cpp

```diff
@@ -12,19 +12,470 @@
 static_assert(is_nothrow_moveable<value>::value, "bsoncxx::v1::types::value must be nothrow moveable");
 
+class alignas(BSONCXX_PRIVATE_MAX_ALIGN_T) value::impl {
+    friend value::internal;
+
+   public:
+    enum : std::size_t {
+        _padding_size = sizeof(value::_storage) // Total reserved.
+                        - sizeof(bson_value_t)  // _value.
+                        - 0u,
+    };
+
+    // `_padding_size == 0` given `sizeof(void*) == 8`.
+    static_assert(_padding_size < sizeof(value::_storage), "sizeof(impl) must not exceed reserved storage size");
+
+    bson_value_t _value = {}; // BSON_TYPE_EOD
+
+    ~impl();
+
+    impl(impl&& other) noexcept {
+        _value = other._value; // Ownership transfer.
+        other._value = {BSON_TYPE_NULL, {}, {}};
+    }
+
+    impl& operator=(impl&& other) noexcept {
+        if (&other != this) {
+            bson_value_destroy(&_value);
+            _value = other._value; // Ownership transfer.
+            other._value = {BSON_TYPE_NULL, {}, {}};
+        }
+        return *this;
+    }
+
+    impl(impl const& other) {
+        bson_value_copy(&other._value, &_value);
+    }
+
+    impl& operator=(impl const& other) {
+        if (&other != this) {
+            bson_value_destroy(&_value);
+            bson_value_copy(&other._value, &_value);
+        }
+        return *this;
+    }
+
+    impl() = default;
+
+    auto t() const -> decltype((_value.value_type)) {
+        return _value.value_type;
+    }
+
+    auto t() -> decltype((_value.value_type)) {
+        return _value.value_type;
+    }
+
+    auto v() const -> decltype((_value.value)) {
+        return _value.value;
+    }
+
+    auto v() -> decltype((_value.value)) {
+        return _value.value;
+    }
+};
+
+value::impl::~impl() {
+    static_assert(is_semiregular<impl>::value, "bsoncxx::v1::types::value::impl must be semiregular");
+    static_assert(is_nothrow_moveable<impl>::value, "bsoncxx::v1::types::value::impl must be nothrow moveable");
+
+    static_assert(sizeof(value::_storage) >= sizeof(value::impl), "insufficient size");
+    static_assert(alignof(value) >= alignof(value::impl), "insufficient alignment");
+
+    bson_value_destroy(&_value);
+}
+
+value::~value() {
+    internal::impl(this)->~impl();
+}
+
+value::value(value&& other) noexcept {
+    new (internal::impl(this)) impl{std::move(internal::impl(other))};
+}
+
+value& value::operator=(value&& other) noexcept {
+    *internal::impl(this) = std::move(internal::impl(other));
+    return *this;
+}
+
+value::value(value const& other) {
+    new (internal::impl(this)) impl{internal::impl(other)};
+}
+
+value& value::operator=(value const& other) {
+    *internal::impl(this) = internal::impl(other);
+    return *this;
+}
+
+namespace {
+
+// For backward compatibility, do not prematurely truncate strings in bsoncxx API. Instead, defer handling of potential
+// embedded null bytes to the bson library.
```

Comment:
This comment documents an observation that v_noabi's [`make_copy_for_libbson`](https://github.com/mongodb/mongo-cxx-driver/blob/ad622035ec9c9ab352b42449d5fcbf426e8782e9/src/bsoncxx/lib/bsoncxx/private/convert.hh#L31) (which in hindsight should really be moved out of `private` and into `v_noabi` along with a few more current `private` headers) uses `memcpy` rather than `str(n)cpy` to create owning BSON strings. The v1 implementation preserves this behavior, avoiding truncation and deferring null byte validation to the bson library.